File: packages/protocol/src/types/preKeyBundle.ts
import { z } from "zod";

export const PreKeyBundleSchema = z.object({
  type: z.literal("prekey_bundle"),
  version: z.literal(1),
  bundle_id: z.string().uuid(),
  identity_id: z.string(),
  device_id: z.string().uuid(),
  device_pubkey: z.string(),
  signed_prekey: z.object({
    key_id: z.string(),
    public_key: z.string(),
    created_at_ms: z.number(),
    expires_at_ms: z.number(),
    signature: z.string(),
  }),
  one_time_prekeys: z.array(
    z.object({
      key_id: z.string(),
      public_key: z.string(),
    }),
  ),
  created_at_ms: z.number(),
  signature: z.string(),
});

export type PreKeyBundle = z.infer<typeof PreKeyBundleSchema>;

File: apps/relay/internal/model/prekey_bundle.go
package model

type PreKeyBundle struct {
	Type           string          `json:"type"`
	Version        int             `json:"version"`
	BundleID       string          `json:"bundle_id"`
	IdentityID     string          `json:"identity_id"`
	DeviceID       string          `json:"device_id"`
	DevicePubKey   string          `json:"device_pubkey"`
	SignedPreKey   SignedPreKey    `json:"signed_prekey"`
	OneTimePreKeys []OneTimePreKey `json:"one_time_prekeys"`
	CreatedAtMs    int64           `json:"created_at_ms"`
	Signature      string          `json:"signature"`
}

type SignedPreKey struct {
	KeyID       string `json:"key_id"`
	PublicKey   string `json:"public_key"`
	CreatedAtMs int64  `json:"created_at_ms"`
	ExpiresAtMs int64  `json:"expires_at_ms"`
	Signature   string `json:"signature"`
}

type OneTimePreKey struct {
	KeyID     string `json:"key_id"`
	PublicKey string `json:"public_key"`
}
